<?xml version="1.0"?>
<robot name="lamp">

  <!-- *************************************************** -->
  <!-- ******************** Kinematics ******************* -->
  <!-- *************************************************** -->

  <!-- Base Link -->
  <link name="base">
    <visual>
      <geometry>
        <mesh filename="package://pixar/meshes/base.dae" scale="0.01 0.01 0.01"/>
      </geometry>
    </visual>
  </link>

  <!-- Joint between world and link0 -->
  <joint name="theta1" type="continuous">
    <parent link="base"/>
    <child  link="link0"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <axis   xyz="0 0 1"/>
  </joint>

  <!-- Link 0 -->
  <link name="link0">
    <visual>
      <geometry>
        <mesh filename="package://pixar/meshes/link.dae" scale="10 10 10"/>
      </geometry>
    </visual>
  </link>

  <!-- Joint between link0 and link1 -->
  <joint name="theta2" type="continuous">
    <parent link="link0"/>
    <child  link="link1"/>
    <origin xyz="0 0 0.5" rpy="0 0 0"/>
    <axis   xyz="0 1 0"/>
  </joint>

  <!-- Link 1 -->
  <link name="link1">
    <visual>
      <geometry>
        <mesh filename="package://pixar/meshes/link.dae" scale="10 10 10"/>
      </geometry>
    </visual>
  </link>

  <!-- Joint between link1 and link2 -->
  <joint name="theta3" type="continuous">
    <parent link="link1"/>
    <child  link="link2"/>
    <origin xyz="0 0 0.5" rpy="0 0 0"/>
    <axis   xyz="0 1 0"/>
  </joint>

  <!-- Link 2 -->
  <link name="link2">
    <visual>
      <geometry>
        <mesh filename="package://pixar/meshes/link.dae" scale="10 10 10"/>
      </geometry>
    </visual>
  </link>

  <!-- Joint theta4: Rotation around Y-axis -->
  <joint name="theta4" type="continuous">
    <parent link="link2"/>
    <child link="head_theta4"/>
    <origin xyz="0 0 0.5" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
  </joint>

  <!-- Intermediate Link for Universal Joint -->
  <link name="head_theta4">
    <visual>
      <geometry>
        <box size="0 0 0"/> <!-- Zero-sized link acting as a connector -->
      </geometry>
    </visual>
  </link>

  <!-- Joint theta5: Rotation around X-axis -->
  <joint name="theta5" type="continuous">
    <parent link="head_theta4"/>
    <child link="head"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <axis xyz="1 0 0"/>
  </joint>

  <!-- Head Link -->
  <link name="head">
    <visual>
      <geometry>
        <mesh filename="package://pixar/meshes/head.dae" scale="10 10 10"/>
      </geometry>
    </visual>
  </link>

</robot>